### DISCUSSION TITLE
Methods and algorithm for determining isFlying()

### DISCUSSION THREAD

**Discussion by sensei-hacker** [iNavFlight/inav#8745] [General]
It was suggested I open a discussion about this, and particularly get feedback from @breadoven 

I'm wanting to make the isFlying () work as well as possible. The instant purpose is to allow re-arm for a few seconds after an accidental in-flight disarm.

To allow re-arming after disarming in flight, we need to detect flight.

In this PR:

https://github.com/iNavFlight/inav/pull/8736#issuecomment-1407187440

I suggested a general approach something like this:

<code>
If landing detected {
 likely_flying = 0
}

while armed and throttle high and speed high {
  likely_flying++;
}

while throttle low or speed is low {
 likely_flying--;
}

while disarmed {
   likely_flying -= 10 # Quickly reduce likelihood we're still airborne
}

</code>

**Comment by MrD-RC** [comment]
I think the 2nd one (--) is a bit too vague. It probably needs to be more platform specific too. 

For example. A motor glider could have zero throttle, not be moving fast, yet could be flying for hours like that. I don't think throttle position should be used to determine if a fixed wing is flying or not. Even with non-gliders, I've flown with zero throttle for a minute or two. Though in those instances, there was a 3D speed that would have been picked up. 

A rotor craft could only have zero throttle for a couple of seconds before it hits the ground.

My thoughts are actually moving away from in-flight _is still flying_ detection, to having more definitive _has started flying_ and _has landed_ detection. If those two work well enough, _is still flying_ is just a case of `has started flying` and `!has landed`. @breadoven has done some great work on landing detection.

For start flight detection, we have an excellent trigger for some fixed wing already in autolaunch. If that completes, you're flying. If you don't use autolaunch, for fixed wing, a `3D speed > 900 cm/s` and an `altitude > 20 m` and a `flight time > 10 s` may be a fair starting point for launch detection. With rotor craft, maybe looking at the airmode triggers would offer some solutions for if the craft is flying.

**Reply by sensei-hacker** [reply]
I realize now I wasn't clear. I was referring only to replacing the current check. Currently it checks if throttle is greater than cruise throttle. That doesn't mean we shouldn't look at anything else, such as velocity or altitude. Yes, if the vehicle is armed, traveling at speed, it's 150 meters above home, and the pilot is moving the sticks, you're probably flying.

Thinking specifically about how we consider throttle, if cruise throttle is 1700 and I've been using 1650, I'm probably flying! *When looking at the throttle component* as evidence, I think the  relevant values of throttle are:

A. Very low (if throttle has been 1050 for 10 minutes, we may not be flying anymore)

B. High (close to cruise throttle, enough to cause flight, a strong hint that we're likely flying)

C. In-between (doesn't provide evidence of whether we're flying or not)

You mentioned you might glide for several seconds. The current isFlying() code, checking for greater than cruise throttle, returns false if you glide for a couple milliseconds, or even descend by setting throttle slightly lower than cruise.

I'm suggesting that *for the throttle check portion*, if throttle has been close to cruise for a while, that's not evidence you aren't flying - it's evidence that you probably ARE flying.

**Reply by MrD-RC** [reply]
More like you could be flying for 10 minutes if the throttle is off completely. There are pilots who thermal and get lift off slopes using INAV. Even with my Baby AR Wing Pro. I was hovering at zero throttle above some trees for a couple of minutes. Due to the wind conditions on that day. 

I’d personally only look at throttle for the initial launch. I don’t think throttle can be a reliable indicator of flight on all platforms.

**Comment by MrD-RC** [comment]
This may be a useful link for this discussion, though it's old now. https://github.com/iNavFlight/inav/issues/2331
Also https://github.com/iNavFlight/inav/issues/6789
